### PR TITLE
[no ticket][risk=no] Puppeteer test log improvement

### PR DESCRIPTION
**Pull request by aweng98**
General Improvement: 
- Clean up code
- Add comments
- Create lesser number of test logs.
      > Only save logs when test fails. Log of passed tests are not saved. This change should makes easier to find failed test log 
         in CircleCI.

### CODE REVIEW

File: e2e/jest.test-setup.ts [hunk 1]
```diff
@@ -42,7 +42,14 @@ beforeEach(async () => {
       });
   });
 
-  /** Emitted when a request fails. */
+  // Emitted when a request fails: 4xx..5xx status codes
+  page.on('requestfailed', async (request) => {
+    if (canLogResponse(request)) {
+      await logError(request);
+    }
+  });
```
Comment (aweng98) [inline e2e/jest.test-setup.ts:50]: Log failed request.

File: e2e/jest.test-setup.ts [hunk 2]
```diff
@@ -88,13 +95,17 @@ beforeEach(async () => {
...
     const title = await getPageTitle();
     try {
-      const args = await Promise.all(message.args().map((a) => describeJsHandle(a)));
+      const args = await Promise.all(message.args().map((jsHandle) => describeJsHandle(jsHandle)));
+      let concatenatedText = '';
+      for (let i = 0; i < args.length; ++i) {
+        if (args[i] !== undefined) {
+          concatenatedText += `${args[i]}\n`;
+        }
+      }
```
Comment (aweng98) [inline e2e/jest.test-setup.ts:104]: Slight change in how we retrieve browser Console messages. There were some Console messages are not logged by the old way.

File: e2e/libs/jest-reporter.js [hunk 3]
```diff
@@ -3,37 +3,81 @@ const path = require('path');
 const winston = require('winston');
 
 module.exports = class JestReporter {
+  testName;
+  testLogName;
+  logger;
```
Comment (aweng98) [inline e2e/libs/jest-reporter.js:8]: Extracted from `onTestResult` to make them available to all functions.

File: e2e/libs/jest-reporter.js [hunk 4]
```diff
@@ -3,37 +3,81 @@ const path = require('path');
...
   onTestResult(_testRunConfig, testResult, _runResults) {
-    const testName = path.parse(testResult.testFilePath).name;
-    let testLogName = `${this.logDir}/${testName}.log`;
-    testResult.testResults.forEach((result) => {
-      const status = result.status;
-      if (status === 'failed') {
-        testLogName = `${this.logDir}/${testName}-FAILED.log`;
+    const hasFailure = testResult.testResults.some((result) => {
+      return result.status === 'failed';
+    });
```
Comment (aweng98) [inline e2e/libs/jest-reporter.js:29]: Extracted out to a new function: `createLogger(fileName)`.

File: e2e/puppeteer-custom-environment.ts [hunk 5]
```diff
@@ -41,10 +41,9 @@ class PuppeteerCustomEnvironment extends PuppeteerEnvironment {
           }
           await fs.ensureDir(this.screenshotDir);
           await fs.ensureDir(this.htmlDir);
-          const timestamp = this.localDateTimeString();
-          const screenshotFile = `${this.screenshotDir}/${testName}_${timestamp}.png`;
+          const screenshotFile = `${this.screenshotDir}/${testName}.png`;
```
Comment (aweng98) [inline e2e/puppeteer-custom-environment.ts:44]: Removed timestamp string as it made file name too long.

File: e2e/libs/jest-reporter.js (hunk 4 above)
Comment (aweng98) [inline e2e/libs/jest-reporter.js:29]: Changed `.forEach` to `.some`.

File: e2e/libs/jest-reporter.js [hunk 6]
```diff
@@ -3,37 +3,81 @@ const path = require('path');
...
-  }
-
-  timeNow() {
-    return new Date().toLocaleString('en-US', {
-      timeZone: 'America/New_York',
-      hour12: false
-    });
+    this.logDir = options.outputdir || 'logs/jest';
+    this.summaryFile = options.filename || 'test-results-summary.json';
   }
```
Comment (aweng98) [inline e2e/libs/jest-reporter.js:16]: Moved to the end of class.

File: e2e/jest.test-setup.ts [hunk 7]
```diff
@@ -212,7 +218,19 @@ const notRedirectRequest = (request: Request): boolean => {
 };
 
 const getResponseText = async (request: Request): Promise<string> => {
-  return (await request.response().buffer()).toString();
+  const REDIRECT_CODE_START = 299;
+  const REDIRECT_CODE_END = 400;
+  const NO_CONTENT_RESPONSE_CODE = 204;
+  const response = request.response();
+  // Log response if response it's not a redirect or no-content
+  const status = response && response.status() ? response.status() : null;
```
Comment (ericsong) [inline e2e/jest.test-setup.ts:226]: I think `response && response.status()` should be basically equivalent to `response && response.status() ? response.status() : null`?

it'll return response.status() if its non falsy

File: e2e/jest.test-setup.ts [hunk 8]
```diff
@@ -212,7 +218,19 @@ const notRedirectRequest = (request: Request): boolean => {
...
+  if (
+    status &&
+    !(status > REDIRECT_CODE_START && status < REDIRECT_CODE_END) &&
```
Comment (ericsong) [inline e2e/jest.test-setup.ts:229]: I think it's a little easier to read values that are inclusive of the start and end. ex >=300 and <= 400.

Also, I think the redirect codes only range from 300-308.

File: e2e/jest.test-setup.ts [hunk 9]
```diff
@@ -212,7 +218,19 @@ const notRedirectRequest = (request: Request): boolean => {
...
+    !(status === NO_CONTENT_RESPONSE_CODE)
```
Comment (ericsong) [inline e2e/jest.test-setup.ts:230]: `status !== NO_CONTENT_RESPONSE_CODE` is easier to read IMO

File: e2e/jest.test-setup.ts [hunk 10]
```diff
@@ -88,13 +95,17 @@ beforeEach(async () => {
     if (!message.args().length) return;
     const title = await getPageTitle();
     try {
-      const args = await Promise.all(message.args().map((a) => describeJsHandle(a)));
+      const args = await Promise.all(message.args().map((jsHandle) => describeJsHandle(jsHandle)));
+      let concatenatedText = '';
+      for (let i = 0; i < args.length; ++i) {
+        if (args[i] !== undefined) {
+          concatenatedText += `${args[i]}\n`;
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/jest.test-setup.ts:102]: 👍🏻 Updated. I haven't noticed anything wrong after replacing code.

File: e2e/jest.test-setup.ts (hunk 7 above)
Comment (aweng98) [reply e2e/jest.test-setup.ts:226]: 👍🏻 Updated.

File: e2e/jest.test-setup.ts (hunk 9 above)
Comment (aweng98) [reply e2e/jest.test-setup.ts:230]: Updated.

File: e2e/jest.test-setup.ts (hunk 8 above)
Comment (aweng98) [reply e2e/jest.test-setup.ts:229]: Updated.

**Review by ericsong — APPROVED** [review]

**Pull request merged**